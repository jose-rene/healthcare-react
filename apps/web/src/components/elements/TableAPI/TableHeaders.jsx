import React from "react";
import { mapTypeToClass } from "../../../helpers/table";
import Icon from "../Icon";

const TableHeaders = ({
    headers = [],
    defaultSortColumn = "na",
    searchObj: { sortDirection = defaultSortColumn, sortColumn = "" },
    onChange,
}) => {
    const handleOnChange = async (columnMap) => {
        if (columnMap === sortColumn) {
            const direction = sortDirection !== "asc" ? "asc" : "desc";

            onChange({ sortDirection: direction, page: 1 });
        } else {
            await onChange({
                sortColumn: columnMap,
                sortDirection: "asc",
                page: 1,
            });
        }
    };

    const sortArrow = (columnMap) => {
        if (columnMap !== sortColumn) {
            return <Icon size="1x" icon="sort" />;
        }

        switch (sortDirection) {
            case "asc":
                return <Icon size="1x" icon="sort-up" />;
            case "desc":
                return <Icon size="1x" icon="sort-down" />;
            default:
                return <Icon size="1x" icon="sort" />;
        }
    };

    return (
        <thead>
            <tr>
                {headers.map((head, index) => {
                    const {
                        columnMap,
                        type,
                        label = columnMap,
                        disableSortBy = false,
                    } = head;

                    if (disableSortBy) {
                        return (
                            <th
                                key={`table-th-${index}`}
                                className={`${mapTypeToClass(type)}`}
                            >
                                {label}
                            </th>
                        );
                    }

                    return (
                        columnMap && (
                            <th
                                key={`table-th-${index}`}
                                className={`c-pointer ${mapTypeToClass(type)}`}
                                onClick={() => handleOnChange(columnMap)}
                            >
                                {label}{" "}
                                <span className="float-right ms-1">
                                    {sortArrow(columnMap)}
                                </span>
                            </th>
                        )
                    );
                })}
            </tr>
        </thead>
    );
};

export default TableHeaders;
