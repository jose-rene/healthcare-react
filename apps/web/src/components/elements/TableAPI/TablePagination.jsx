import React, { useMemo } from "react";
import { Pagination } from "react-bootstrap";
import useWindowSize from "../../../hooks/useWindowSize";
import Select from "../../inputs/Select";
import { range } from "lodash";

const TablePagination = ({totalPages, onChange, lastPage, searchObj}) => {
    const {width} = useWindowSize();
    const {
        page = 1,
        perPage = 50,
    } = searchObj;

    const items = useMemo(() => {
        if(totalPages <= 1){
            return null
        }

        const active = page;
        const _items = [
            <Pagination.First
                key={`pagination-item-first`}
                active={active <= 1}
                onClick={() => onChange(1)}
            />,
            <Pagination.Prev
                key={`pagination-item-prev`}
                active={active == 1}
                onClick={() => onChange(active-1)}
            />
        ];

        for (let number = 1; number <= totalPages; number++) {
            _items.push(
                <Pagination.Item
                    key={`pagination-item-${number}`}
                    active={number === active}
                    onClick={() => onChange(number)}
                >
                    {number}
                </Pagination.Item>
            );
        }

        _items.push(
            <Pagination.Next
                key={`pagination-item-next`}
                active={active == lastPage}
                onClick={() => onChange(active+1)}
            />
        );
        _items.push(
            <Pagination.Last
                key={`pagination-item-last`}
                active={active == lastPage}
                onClick={() => onChange(lastPage)}
            />
        )

        return _items;
    }, [page, perPage]);

    if (totalPages > 10 || width < 480) {
        return (
            <Select
                label="Page"
                inlineLabel
                name="page"
                value={page}
                options={range(1, lastPage).map(n => ({
                    id: n,
                    value: n,
                    title: n,
                }))}
                onChange={({target: {value}}) => onChange(value)}
            />
        );
    }

    return (
        <Pagination>
            {items}
        </Pagination>
    );
}

export default TablePagination;
