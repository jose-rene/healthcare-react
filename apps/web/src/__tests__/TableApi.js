import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";
import TablePagination from "../components/elements/TableAPI/TablePagination";
import TableHeaders from "../components/elements/TableAPI/TableHeaders";

// bug with coverage https://github.com/facebook/create-react-app/issues/8689

describe("<TableAPI />", () => {
    it("renders pagination with a select", () => {
        const props = {
            onChange: jest.fn(),
            totalPages: 21,
            lastPage: 2,
            searchObj: {
                page: 1,
                perPage: 20,
            },
        };

        const pagination = renderer
            .create(<TablePagination {...props} />)
            .toJSON();
        expect(pagination).toMatchSnapshot();
    });

    it("renders regular pagination", () => {
        const props = {
            onChange: jest.fn(),
            totalPages: 9,
            lastPage: 2,
            searchObj: {
                page: 1,
                perPage: 20,
            },
        };

        const pagination = renderer
            .create(<TablePagination {...props} />)
            .toJSON();
        expect(pagination).toMatchSnapshot();
    });

    it("renders sortable headers", () => {
        const props = {
            onChange: jest.fn(),
            headers: [
                {
                    label: "Title",
                    columnMap: "roles.0.title", // name.title
                    disableSortBy: true,
                },
                {
                    label: "First Name",
                    columnMap: "first_name",
                },
                {
                    label: "Last Name",
                    columnMap: "last_name",
                },
                {
                    label: "Email",
                    columnMap: "email",
                    // disableSortBy: true,
                },
                {
                    label: "City",
                    columnMap: "address.city",
                },
            ],
            searchObj: {},
        };

        const tableHeader = renderer
            .create(<TableHeaders {...props} />)
            .toJSON();
        expect(tableHeader).toMatchSnapshot();
    });
});
