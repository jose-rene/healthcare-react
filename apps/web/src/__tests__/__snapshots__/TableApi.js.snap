// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableAPI /> renders pagination with a select 1`] = `
<div
  className="mb-3 form-floating form-group d-flex"
>
  <div
    className="form-floating"
    id="page"
  >
    <select
      className="col   form-select"
      id="page"
      name="page"
      onChange={[Function]}
      value={1}
    >
      <option
        value={1}
      >
        1
      </option>
    </select>
    <label>
      Page
    </label>
  </div>
</div>
`;

exports[`<TableAPI /> renders regular pagination 1`] = `
<ul
  className="pagination"
>
  <li
    className="page-item active"
  >
    <span
      className="page-link"
      disabled={false}
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
      >
        «
      </span>
      <span
        className="visually-hidden"
      >
        First
      </span>
      <span
        className="visually-hidden"
      >
        (current)
      </span>
    </span>
  </li>
  <li
    className="page-item active"
  >
    <span
      className="page-link"
      disabled={false}
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
      >
        ‹
      </span>
      <span
        className="visually-hidden"
      >
        Previous
      </span>
      <span
        className="visually-hidden"
      >
        (current)
      </span>
    </span>
  </li>
  <li
    className="page-item active"
  >
    <span
      className="page-link"
      disabled={false}
      onClick={[Function]}
    >
      1
      <span
        className="visually-hidden"
      >
        (current)
      </span>
    </span>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      2
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      3
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      4
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      5
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      6
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      7
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      8
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      9
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        aria-hidden="true"
      >
        ›
      </span>
      <span
        className="visually-hidden"
      >
        Next
      </span>
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        aria-hidden="true"
      >
        »
      </span>
      <span
        className="visually-hidden"
      >
        Last
      </span>
    </a>
  </li>
</ul>
`;

exports[`<TableAPI /> renders sortable headers 1`] = `
<thead>
  <tr>
    <th
      className="text-left align-middle"
    >
      Title
    </th>
    <th
      className="c-pointer text-left align-middle"
      onClick={[Function]}
    >
      First Name
       
      <span
        className="float-right ms-1"
      >
        <i
          className="icon fas fa-sort   fa-1x "
        />
      </span>
    </th>
    <th
      className="c-pointer text-left align-middle"
      onClick={[Function]}
    >
      Last Name
       
      <span
        className="float-right ms-1"
      >
        <i
          className="icon fas fa-sort   fa-1x "
        />
      </span>
    </th>
    <th
      className="c-pointer text-left align-middle"
      onClick={[Function]}
    >
      Email
       
      <span
        className="float-right ms-1"
      >
        <i
          className="icon fas fa-sort   fa-1x "
        />
      </span>
    </th>
    <th
      className="c-pointer text-left align-middle"
      onClick={[Function]}
    >
      City
       
      <span
        className="float-right ms-1"
      >
        <i
          className="icon fas fa-sort   fa-1x "
        />
      </span>
    </th>
  </tr>
</thead>
`;
